impeller/renderer/backend/vulkan/fence_waiter_vk.cc

```diff
@@ -101,76 +95,94 @@ void FenceWaiterVK::Main() {
     // critical section. Copy the array of entries and immediately unlock the
     // mutex.
     WaitSet wait_set = wait_set_;
-
-    const auto terminate = terminate_;
-
     lock.unlock();
 
+    const auto terminate = terminate_;
     if (terminate) {
+      WaitUntilEmpty();
       break;
     }
 
-    // Check if the context had died in the meantime.
-    auto device_holder = device_holder_.lock();
-    if (!device_holder) {
+    if (!Wait(wait_set)) {
       break;
     }
+  }
+}
 
-    const auto& device = device_holder->GetDevice();
+void FenceWaiterVK::WaitUntilEmpty() {
+  // Note, there is no lock because once terminate_ is set to true, no other
+  // fence can be added to the wait set. Just in case, here's a FML_DCHECK:
+  FML_DCHECK(terminate_) << "Fence waiter must be terminated.";
+  while (!wait_set_.empty() && Wait(wait_set_)) {
```

Comment:
This line is passing a reference to the `wait_set_` member to `Wait`.  So `Wait`'s `wait_set` argument will be an alias for the `wait_set_` member.

`Wait` mutates both `wait_set` and `wait_set_`, which will make this error prone.

This could be simplified by removing the `wait_set` parameter and having `Wait` do the snapshotting of `wait_set_` internally.

Reply:
Nice suggestion, I agree the previous code was very complicated. Done.